lib/options.nix

```diff
@@ -187,6 +189,218 @@ rec {
   */
   mkPackageOptionMD = mkPackageOption;
 
+  /* Creates a suggestion based resource namespace allocator.
+
+    It allows someone to define a value space, like a port interval, so other
+    modules can "ask" for a value and the module will check for missing values,
+    conflicts and invalid candidates and suggest values for these cases that
+    must be explicitly set.
+
+    See RFC 0159 for more specification details.
+
+    Example:
+     with lib; mkAllocatorModule {
+      # dontThrow = true;
+      keyPath = "networking.ports";
+
+      valueKey = "port";
+      valueType = types.port;
+      cfg = config.networking.ports;
+      description = "Build time port allocations for services that are only used internally";
+      enableDescription = name: "Enable automatic port allocation for service ${name}";
+      valueDescription = name: "Allocated port for service ${name}";
+
+      firstValue = 49151;
+      succFunc = x: x - 1;
+      valueLiteral = toString;
+      validateFunc = x: (types.port.check x) && (x > 1024);
+      example = literalExpression ''{
+        app = {
+          enable = true;
+          port = 42069; # guided
+        };
+      }'';
+    }
+    => { _type = "option"; apply = «lambda @ /nix/store/8dq1grzr2g9ghh8nrwh15mn65pnp9qb1-source/lib/options.nix:309:13»; default = { ... }; description = "Build time port allocations for services that are only used internally"; example = { ... }; internal = null; relatedPackages = null; type = { ... }; visible = null; }
+  */
+  mkAllocatorModule = {
+    # <!-- begin of passthru parameters to the outer mkOption -->
+    # `example` parameter passed to the outer mkOption
+    example ? null,
+    # `internal` parameter passed to the outer mkOption
+    internal ? null,
+    # `relatedPackages` parameter passed to the outer mkOption
+    relatedPackages ? null,
+    # `visible` parameter passed to the outer mkOption
+    visible ? null,
+    # `description` parameter passed to the outer mkOption
+    description ? null,
+    # <!-- end of passthru parameters to the outer mkOption -->
```

Comment:
This is a lot of duplication. Couldn't the allocator be returned as a type instead?
Instead of `apply`, you could post-compose to the `merge` function.